store/src/java/com/zimbra/cs/filter/SieveToSoap.java

```diff
@@ -142,21 +159,40 @@ protected void visitIfControl(Node ruleNode, VisitPhase phase, RuleProperties pr
                 addAction(action);
                 actionVariables = null;
             }
+
+            if (numOfIfProcessingStarted == numOfIfProcessingDone) {
+                // If all the nested 'if' is completed, clean up the rule objects
+                currentRule = null;
+                currentNestedRule = null;
+            }
+            visitRule(ruleNode, VisitPhase.end, props);
             return;
         }
         numOfIfProcessingStarted++;   // number of if for which process is started.
+        visitRule(ruleNode, VisitPhase.begin, props);
     }
 
     private boolean isNestedRule(){
         // in non nested case, only one process is started but not done.
-        return !(numOfIfProcessingStarted == numOfIfProcessingDone + 1);
+        if (numOfIfProcessingStarted == numOfIfProcessingDone + 1) {
+            return false;
+        }
+        // if an action is specified outside of the 'if' clause,
+        // the number of Started becomes equal to that of Done.
+        if (numOfIfProcessingStarted == numOfIfProcessingDone) {
```

Comment:
As described in the comment, this 'if' statement is necessary.  And also good for the visibility.